File: Sprint1/Sprint1_ProblemAnalysis/resources/wasteservice/WasteService.kt
package wasteservice

import java.util.*

class WasteService {
    // The content of the containers (to handle the requests and be able to respond instantly if its a loadrejecteed)
    private var preStorage = EnumMap<WasteType, Double>(WasteType.values().associateWith { 0.0 })
    private var storage = EnumMap<WasteType, Double>(WasteType.values().associateWith { 0.0 })
    private var storageCapacity = EnumMap<WasteType, Double>(WasteType.values().associateWith { WSconstants.DEFAULT_MAX })

    // fun canStore // check if there is enough space
    // fun store // store the waste

    fun setCapacity(wType: WasteType, wCapacity: Double) {
        storageCapacity[wType] = wCapacity;
    }
    fun getCurrentPreStorageString(): String {
        var res: String = ""

        for(wt in preStorage.keys) {
            res += "${wt.name}: ${preStorage[wt]}\n"
        }

        return res
    }
    fun getCurrentStorageString(): String {
        var res: String = ""

        for(wt in storage.keys) {
            res += "${wt.name}: ${storage[wt]}\n"
        }

        return res
    }
    fun getFullStatusString(): String {
        var res: String = ""
        for(wt in storage.keys) {
            res += "${wt.name}: ${preStorage[wt]}/${storage[wt]}/${storageCapacity[wt]}\n"
        }

        return res
    }
    fun canPreStore(wType: WasteType, wWeight: Double): Boolean {
        if(!preStorage.keys.contains(wType))
            return false
        if(wWeight <= 0.0)
            return false
        if(preStorage[wType]!! + wWeight <= storageCapacity[wType]!!)
            return true
        return false;
    }
    fun canStore(wType: WasteType, wWeight: Double): Boolean {
        if(!storage.keys.contains(wType))
            return false
        if(wWeight <= 0.0)
            return false
        if(storage[wType]!! + wWeight <= storageCapacity[wType]!!)
            return true
        return false;
    }
    fun addToPreStorage(wType: WasteType, wWeight: Double) {
        if(canPreStore(wType, wWeight)) {
            var oldLoad = preStorage[wType]
            preStorage[wType] = oldLoad!! + wWeight
        }
    }
    fun addToStorage(wType: WasteType, wWeight: Double) {
        if(canStore(wType, wWeight)){
            var oldLoad = storage[wType]
            storage[wType] = oldLoad!! + wWeight
        }
    }
}

File: Sprint1/Sprint1_Project/sprint1_smart_device/lib/model/waste_service/waste_service.dart
const double defaultMax = 100.0;

class WasteService {
  Map<String, double> _preStorage = {};
  Map<String, double> _storage = {};
  Map<String, double> _storageCapacity = {};

  // canStore
  WasteService(List<String> wasteTypes, {double? capacity}) {
    for (String s in wasteTypes) {
      _preStorage[s] = 0.0;
      _storage[s] = 0.0;
      _storageCapacity[s] = capacity ?? defaultMax;
    }
  }

  String getTypesListString(String separator) {
    String res = "";

    for (String s in _storage.keys) {
      res += s;
      if (s != _storage.keys.last) {
        res += separator;
      }
    }

    return res;
  }

  bool canPreStore(String type, double weight) {
    if (!_preStorage.containsKey(type)) {
      return false;
    }
    return _preStorage[type]! + weight <= _storageCapacity[type]!;
  }

  bool canStore(String type, double weight) {
    if (!_storage.containsKey(type)) {
      return false;
    }
    return _storage[type]! + weight <= _storageCapacity[type]!;
  }

  // pre deposit
  void addToPreStorage(String type, double weight) {
    if (canPreStore(type, weight)) {
      _preStorage[type] = _preStorage[type]! + weight;
    }
  }

  // deposit
  void addToStorage(String type, double weight) {
    if (canStore(type, weight)) {
      _storage[type] = _storage[type]! + weight;
    }
  }

  String getStatusString(String type) {
    String res = "";

    if (_storage.containsKey(type)) {
      res =
          "$type: ${_preStorage[type]}/${_storage[type]}/${_storageCapacity[type]} KG\n";
    }

    return res;
  }

  String getFullStatusString() {
    String res = "";

    for (String type in _preStorage.keys) {
      res +=
          "$type: ${_preStorage[type]}/${_storage[type]}/${_storageCapacity[type]} KG\n";
    }

    return res;
  }
}
